plans/phase-1/api-design/multigres-operator-api-v1alpha1-design.md

```diff
@@ -0,0 +1,1121 @@
+  * **Override Chain:** All components use the following 4-level precedence chain for configuration:
+    1.  **Component-Level Definition:** An inline `spec` (e.g., `managedSpec`) or an explicit `templateRef` on the component itself.
+    2.  **Cluster-Level Default:** The corresponding template defined in `spec.templateDefaults` (e.g., `templateDefaults.coreTemplate` or `templateDefaults.cellTemplate`). 
+    3.  **Namespace-Level Default:** A template of the correct kind (e.g., `CoreTemplate`) named `default` in the same namespace.
+    4.  **Operator Hardcoded Defaults:** A final fallback applied by the operator's admission webhook.
+  * **Atomic Overrides:** To ensure safety, highly interdependent fields are grouped (e.g., `resources`, `storage`). When using `overrides`, you must replace the *entire* group, not just individual sub-fields (e.g., you cannot override just `cpu limit` without also providing `cpu request`).
+  * Images are defined globally to avoid the danger of running multiple incongruent versions at once. This implies the operator handles upgrades.
+  * The `MultigresCluster` does not create its grandchildren directly; for example, shard configuration is passed to the `TableGroup` CR, which then creates its own children `Shard` CRs.
+
+
+```yaml
+apiVersion: multigres.com/v1alpha1
+kind: MultigresCluster
+metadata:
+  name: example-cluster
+  namespace: example
+spec:
+  # ----------------------------------------------------------------
+  # Global Images Cluster Configuration
+  # ----------------------------------------------------------------
+  # Images are defined globally to ensure version consistency across
+  # all cells and shards.
+  # NOTE: Perhaps one day we can template these as well.
+  images:
+    imagePullPolicy: "IfNotPresent"
+    imagePullSecrets:
+      - name: "my-registry-secret"
+    multigateway: "multigres/gateway:latest"
+    multiorch: "multigres/orch:latest"
+    multipooler: "multigres/pooler:latest"
+    multiadmin: "multigres/admin:latest"
+    postgres: "postgres:15.3"
+
+  # ----------------------------------------------------------------
+  # Cluster-Level Template Defaults
+  # ----------------------------------------------------------------
+  # These are the default templates to use for any component that
+  # does not have an inline 'spec' or an explicit 'templateRef'.
+  templateDefaults:
+    coreTemplate: "cluster-wide-core"
+    cellTemplate: "cluster-wide-cell"
+    shardTemplate: "cluster-wide-shard"
+
+  # ----------------------------------------------------------------
+  # Global Components (REVISED)
+  # ----------------------------------------------------------------
+  
+  # Global TopoServer is a singleton. It follows the 4-level override chain.
+  # It supports EITHER a 'managedSpec' OR 'external' OR 'templateRef'.
+  globalTopoServer:
+    # --- OPTION 1: Inline Managed Spec ---
+    managedSpec:
```

Comment:
I don't think I asked this before -- why do we call it `managedSpec`, rather than just `spec`?

Reply:
@rytswd To separated from the external block and show that this is managed. I suppose for consistency we could call it just managed, or rename the external to externalSpec. That may make more sense

Reply:
Corrected on f1afd1f